src/IW/Sync/Update.hs

```diff
@@ -1,40 +1,86 @@
+{-# LANGUAGE MultiWayIf #-}
+
 {- | This module contains functions that are used for updating the database.
 It combines functionality from @IW.Sync.Search@ and @IW.Db@ to fetch the latest
 data and insert it into the database.
 -}
 
 module IW.Sync.Update
-       ( fetchAndUpsertRepos
+       ( syncRepos
        ) where
 
 import Control.Monad.IO.Unlift (MonadUnliftIO)
+import Data.Time (Day (..), addDays)
 import UnliftIO.Async (mapConcurrently_)
 
 import IW.App (WithError)
 import IW.Core.Repo (Repo (..))
 import IW.Db (WithDb, upsertRepos, updateRepoCategories)
 import IW.Effects.Cabal (MonadCabal (..), getCabalCategories)
-import IW.Sync.Search (fetchAllHaskellRepos, fromGitHubRepo)
+import IW.Sync.Search (fetchHaskellReposByDate, fromGitHubRepo)
+
+
+-- | This function fetches repos from the GitHub API within a specified date range,
+-- parses their @.cabal@ files, and upserts them into the database.
+syncRepos
+    :: forall env m.
+       ( MonadCabal m
+       , MonadUnliftIO m
+       , WithDb env m
+       , WithLog env m
+       , WithError m
+       )
+    => Day     -- ^ Oldest day
+    -> Day     -- ^ Most recent day
+    -> Integer -- ^ Interval of days before most recent day
+    -> Int     -- ^ Page
+    -> m ()
+syncRepos oldest recent interval page =
+    if recent == oldest then
+        log I $ "Oldest day reached"
+    else
+        do
+            resCount <- syncReposByDate intervalStart recent page
+            if resCount < 100 then
+                syncRepos oldest nextRecent interval 1
+            else
+                syncRepos oldest recent interval (page + 1)
```

Comment:
I feel like there is a better way to implement this function. It looks like some state machine. Maybe there is a way to analyse arguments and decide what to do next depending on that. This ad-hoc if-then-else logic seems okay for now and it works. So maybe this is okay for now. But I wrote similar code before and it was buggy due to off-by-one errors and various similar arithmetics. So I just want to be more cautious :slightly_smiling_face:

Reply:
Hmmm yes, I agree. I try to avoid if-else logic when possible. I'll come up with something better.

Reply:
I'm still trying to think about how I can analyze the arguments and then make a decision based on that rather than using if-then-else logic.